scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java

```diff
@@ -840,13 +870,46 @@ private void processRecord(Record record, Connection connection) {
 		try {
 			// ensure, that connection is still related to record 
 			// and not changed by processing an other record before 
-			if (!connection.equalsPeerAddress(record.getPeerAddress())) {
+			if (record.getConnectionId() == null && !connection.equalsPeerAddress(record.getPeerAddress())) {
 				LOGGER.warn("Drop record {}, connection changed address {} => {}!", record.getType(),
 						record.getPeerAddress(), connection.getPeerAddress());
 				return;
 			}
+			int epoch = record.getEpoch();
 			LOGGER.trace("Received DTLS record of type [{}], length: {}, [epoche:{},reqn:{}]", 
-					record.getType(), record.getLength(), record.getEpoch(), record.getSequenceNumber());
+					record.getType(), record.getFragmentLength(), epoch, record.getSequenceNumber());
+
+			Integer cidLength = config.getConnectionIdLength();
+			boolean useCid = cidLength != null && cidLength > 0;
+			if (record.getType() == ContentType.TLS12_CID) {
+				// !useCid already dropped in Record.fromByteArray
+				if (epoch == 0) {
+					LOGGER.debug("Discarding TLS_CID record received from peer [{}] during handshake",
+							record.getPeerAddress());
+					return;
+				}
+				DTLSSession session = connection.getSession(epoch);
+				if (session == null && epoch > 0) {
+					// received record of epoch 1 during handshake
+					Handshaker handshaker = connection.getOngoingHandshake();
+					if (handshaker != null && handshaker.isChangeCipherSpecMessageExpected()) {
+						handshaker.addRecordsForDeferredProcessing(record);
+					} else {
+						LOGGER.debug("Discarding TLS_CID record received from peer [{}] without an active session",
+								record.getPeerAddress());
+					}
+					return;
+				} else {
+					record.setSession(session);
+					record.getFragment();
```

Comment:
This PR doesn't change that. 
The new record_type with the "encrypted inner type" just requires, that it is called before that real type is used to determine the record type for processing.

Reply:
> This PR doesn't change that.

I agree. Here this is just more visible as we just call a getter without really use the result.

> The new record_type with the "encrypted inner type" just requires, that it is called before that real type is used to determine the record type for processing.

If we don't want do change the idea (I mean a getter which do some processing), maybe this would be more consistent to just "extract the fragment" on getType() ?

Reply:
If you check, where "getType()" is called, you will see the answer.
The inner/outer is a little shitty, but for now, "it works". 
I added to comment to give a hint about the intention.

Reply:
ok